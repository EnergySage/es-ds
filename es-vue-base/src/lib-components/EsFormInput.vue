<template>
    <div
        v-bind="$attrs"
        :required="required">
        <!-- eslint-disable-next-line vuejs-accessibility/label-has-for -->
        <label
            :for="id"
            class="label font-weight-bold">
            <slot name="label" />
            <span
                v-if="required"
                class="text-danger">
                *
            </span>
        </label>
        <div>
            <b-form-input
                :id="id"
                :type="type"
                :disabled="disabled"
                :state="state"
                v-on="$listeners" />
            <b-form-text v-if="state == null">
                <slot name="message" />
            </b-form-text>
            <b-form-invalid-feedback>
                <slot name="errorMessage" />
            </b-form-invalid-feedback>
            <b-form-valid-feedback>
                <slot name="successMessage" />
            </b-form-valid-feedback>
        </div>
    </div>
</template>

<script lang="js">
import {
    BFormText,
    BFormInput,
    BFormInvalidFeedback,
    BFormValidFeedback,
} from 'bootstrap-vue';

export default {
    name: 'EsFormInput',
    components: {
        BFormText,
        BFormInput,
        BFormInvalidFeedback,
        BFormValidFeedback,
    },
    props: {
        /**
         * Required
         */
        required: {
            type: Boolean,
            default: false,
        },
        /**
         * Disabled
         */
        disabled: {
            type: Boolean,
            default: false,
        },
        /**
         * Type
         */
        type: {
            type: String,
            default: 'text',
            validator: (val) => ['text', 'number'].includes(val),
        },
        /**
         * ID
         * required
         */
        id: {
            type: String,
            required: true,
        },
        /**
         * state
         */
        state: {
            type: [Boolean, null],
            default: null,
        },
    },
};
</script>
