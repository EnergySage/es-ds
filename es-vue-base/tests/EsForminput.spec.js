import { mount } from '@vue/test-utils';
import EsFormInput from '@/src/lib-components/molecules/EsFormInput.vue';
import jestVue from '@/tests/jest.vue.config';

describe('EsFormInput', () => {
    // Basic test; should exist for most components
    test('<EsFormInput />', async () => {
        const wrapper = mount(EsFormInput, {
            ...jestVue,
            slots: {
                label: 'Account Number',
            },
            propsData: {
                id: 'myID',
            },
        });
        const a11y = await axe(wrapper.element);

        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();
        expect(a11y).toHaveNoViolations();
    });

    // Test slots are there
    test('<EsFormInput>slots</EsFormInput>', async () => {
        const wrapper = mount(EsFormInput, {
            ...jestVue,
            slots: {
                label: 'Account Number',
                message: 'Please enter your account number.',
                errorMessage: 'Please enter a valid account number.',
                successMessage: 'Your account number has been submitted successfully.',
            },
            propsData: {
                id: 'myID',

            },
        });
        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();
        expect(wrapper.vm.$slots.label[0].text).toBe('Account Number');
        expect(wrapper.vm.$slots.message[0].text).toBe('Please enter your account number.');
        expect(wrapper.vm.$slots.errorMessage[0].text).toBe('Please enter a valid account number.');
        expect(wrapper.vm.$slots.successMessage[0].text).toBe('Your account number has been submitted successfully.');
    });

    // Test label structure
    test('<EsFormInput <label /> />', async () => {
        const wrapper = mount(EsFormInput, {
            ...jestVue,
            propsData: {
                id: 'myID',
                required: true,
            },
        });

        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();

        const label = wrapper.find('label');
        expect(label.exists()).toBe(true);
        expect(label.find('*').exists()).toBe(true);
    });

    // Test props
    test('<EsFormInput props />', async () => {
        const wrapper = mount(EsFormInput, {
            ...jestVue,
            propsData: {
                id: 'myID',
            },
        });

        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();

        expect(wrapper.props('id')).toBe('myID');
        expect(wrapper.props('state')).toBeNull();
    });
});
