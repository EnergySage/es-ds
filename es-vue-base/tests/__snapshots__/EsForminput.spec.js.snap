// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EsFormInput <EsFormInput /> 1`] = `
"<div><label for=\\"myID\\" class=\\"label font-weight-bold\\">Account Number
    <!----></label>
  <div><input id=\\"myID\\" type=\\"text\\" class=\\"form-control\\"> <small class=\\"form-text text-muted\\"></small>
    <div class=\\"invalid-feedback\\"></div>
    <div class=\\"valid-feedback\\"></div>
  </div>
</div>"
`;

exports[`EsFormInput <EsFormInput <label /> /> 1`] = `
"<div required=\\"required\\"><label for=\\"myID\\" class=\\"label font-weight-bold\\"> <span class=\\"text-danger\\">
            *
        </span></label>
  <div><input id=\\"myID\\" type=\\"text\\" class=\\"form-control\\"> <small class=\\"form-text text-muted\\"></small>
    <div class=\\"invalid-feedback\\"></div>
    <div class=\\"valid-feedback\\"></div>
  </div>
</div>"
`;

exports[`EsFormInput <EsFormInput props /> 1`] = `
"<div><label for=\\"myID\\" class=\\"label font-weight-bold\\">
    <!----></label>
  <div><input id=\\"myID\\" type=\\"text\\" class=\\"form-control\\"> <small class=\\"form-text text-muted\\"></small>
    <div class=\\"invalid-feedback\\"></div>
    <div class=\\"valid-feedback\\"></div>
  </div>
</div>"
`;

exports[`EsFormInput <EsFormInput>slots</EsFormInput> 1`] = `
"<div><label for=\\"myID\\" class=\\"label font-weight-bold\\">Account Number
    <!----></label>
  <div><input id=\\"myID\\" type=\\"text\\" class=\\"form-control\\"> <small class=\\"form-text text-muted\\">Please enter your account number.</small>
    <div class=\\"invalid-feedback\\">Please enter a valid account number.</div>
    <div class=\\"valid-feedback\\">Your account number has been submitted successfully.</div>
  </div>
</div>"
`;
